src/main/java/org/sonar/plugins/checkstyle/metadata/CheckstyleMetadata.java

```diff
@@ -0,0 +1,280 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package org.sonar.plugins.checkstyle.metadata;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.example.ModuleDetails;
+import org.example.ModulePropertyDetails;
+import org.example.ModuleType;
+import org.example.XMLReader;
+import org.sonar.api.rule.RuleStatus;
+import org.sonar.api.server.debt.DebtRemediationFunction;
+import org.sonar.api.server.debt.internal.DefaultDebtRemediationFunction;
+import org.sonar.api.server.rule.RuleParamType;
+import org.sonar.api.server.rule.RulesDefinition;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.ClassPath;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.checks.javadoc.AbstractJavadocCheck;
+
+public class CheckstyleMetadata {
+    private static final String XML_EXTENSION = ".xml";
+    private static final Map<String, String> CHECK_ACCEPTABLE_TOKENS = new HashMap<>();
+    private static final Map<String, RuleParamType> MAP_PARAM_VALUES = new HashMap<>();
+    private final RulesDefinition.NewRepository repository;
+
+    public CheckstyleMetadata(RulesDefinition.NewRepository repository) throws
+            IllegalAccessException, InstantiationException {
+        this.repository = repository;
+        createParamsMap();
+        populateCheckAcceptableTokens();
+    }
+
+    private ModuleDetails loadMeta(String checkName) {
+        final ModuleDetails moduleDetails;
+        try (InputStream inputStream = getClass().getResourceAsStream(checkName + XML_EXTENSION)) {
+            moduleDetails = new XMLReader().read(inputStream, ModuleType.CHECK);
+        }
+        catch (IOException ex) {
+            throw new IllegalStateException("exception occured during loadMeta", ex);
+        }
+        return moduleDetails;
+    }
+
+    public void updateRulesWithMetadata() {
+        repository.rules().forEach(rule -> {
+            final String checkName = rule.key().substring(rule.key().lastIndexOf('.') + 1);
+            // condition required only for PoC
+            if ("EmptyBlockCheck".equals(checkName) || "JavadocPackageCheck".equals(checkName)) {
```

Comment:
are those new checks or checks to be updated? are those "complex" checks covering all scenarios of checks?
i'd like to have a PoC covering as many use cases/permutations as possible. :)

Reply:
I have supplid quite a number of checks to form almost all possible use cases I can think of.